### PR TITLE
Add project-specific pull request template

### LINKED ISSUE #57: Create Pull Request Template
# 💡 Summary #

Create a project-specific pull request template to be used when creating a pull request against the ScubaGear repository. 

## Motivation and context ##

The pull request template ensures that each contribution includes necessary information to better guide developers and contributors through the development/review process so that additions/updates can be made quickly and include code quality checks to prevent bugs.

## Implementation notes ##

Please provide details for implementation, such as:

- Use the existing org PR template as a starting point
- Add new template to `.github/pull_request_template.md`
- Adjust the org default to fit the project-specific process
- Post new template in the repository

## Acceptance criteria ##

How do we know when this work is done?

- [ ] New template format and content has been successfully demoed to the development team and project owners
- [ ] Feedback from the demo has been successfully adjudicated and any changes integrated into template
- [ ] Pull request using the new template has been created for adding the new template to the repository

### PR DESCRIPTION
**Pull request by schrolla**
#  #
  

## 🗣 Description ##

Adds a new template for pull requests to the ScubaGear project repository.  

## 💭 Motivation and context ##

The new template is crafted to better align to the project's current development and release process.  Having additional and appropriate checklist items helps developers and contributors ensure that the pull request contains the necessary information for reviewers and has meet the correct criteria for inclusion into the project.

Closes #57 

## 🧪 Testing ##

The template is basic markdown and was tested using a Markdown rendering engine (VSCode extension).  Additionally, the new template was used as the basis for this pull request itself.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title
- [x] PR is targeting the correct parent branch for merge
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced
      in code comments
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] All relevant type-of-change labels have been added
- [x] All relevant project fields have been set
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR
- [x] Tests have been added and/or modified to cover the changes in this PR
- [x] All new and existing tests pass
- [x] All automated checks pass

## ✅ Pre-merge checklist ##

- [ ] Rebase feature branch against changes from parent branch, as needed
- [ ] Merge coordinator notified via PR comment

## ✅ Post-merge checklist ##

- [ ] Verify that all checks pass on parent branch after merge

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+- [ ] PR is targeting the correct parent branch for merge
+- [ ] Changes are limited to a single goal - *eschew scope creep!*
+- [ ] *All* future TODOs are captured in issues, which are referenced
+      in code comments
+- [ ] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
+- [ ] All relevant type-of-change labels have been added
+- [ ] All relevant project fields have been set
+- [ ] All relevant repo and/or project documentation has been updated
+      to reflect the changes in this PR
+- [ ] Tests have been added and/or modified to cover the changes in this PR
```
Comment (gdasher) [inline .github/pull_request_template.md:46]: Suggest we spell out the types of tests that are expected, here or in comments in the testing section.  I'm thinking:

1) Make sure unit test coverage (with appropriate mocks) is added to cover powershell and rego changes
2) Make sure functional tests are added/updated for major changes to baseline content
3) In addition, ensure smoke tests continue to pass.

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+- [ ] All new and existing tests pass
+- [ ] All automated checks pass
```
Comment (gdasher) [inline .github/pull_request_template.md:48]: provide examples: rego linter, ps static analysis & lint, etc.

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+## ✅ Pre-merge checklist ##
+
+<!-- Remove any of the following that do not apply. -->
+<!-- These boxes should remain unchecked until the pull request has been -->
+<!-- approved. -->
+
+- [ ] Rebase feature branch against changes from parent branch, as needed
+- [ ] Merge coordinator notified via PR comment
```
Comment (gdasher) [inline .github/pull_request_template.md:57]: Might be worth explaining this process in a bit more detail in comments, namely that people are not expected to submit their own PRs and instead have the merge coordinator submit them [though maybe we should consider changing that given the improvements in our testing automation?]

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+- [ ] All new and existing tests pass
+- [ ] All automated checks pass
+
+## ✅ Pre-merge checklist ##
+
+<!-- Remove any of the following that do not apply. -->
+<!-- These boxes should remain unchecked until the pull request has been -->
+<!-- approved. -->
+
+- [ ] Rebase feature branch against changes from parent branch, as needed
```
Comment (gdasher) [inline .github/pull_request_template.md:56]: Can you link to instructions on the appropriate rebasing workflow for people who are less skilled with git, even if they're not specific to this project?

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+## ✅ Pre-approval checklist ##
+
+<!-- Remove any of the following that do not apply. -->
+<!-- Draft PRs should have one or more unchecked boxes. -->
+<!-- If you're unsure about any of these, don't hesitate to ask. -->
+<!-- We're here to help! -->
+
+- [ ] This PR has an informative and human-readable title
+- [ ] PR is targeting the correct parent branch for merge
```
Comment (gdasher) [inline .github/pull_request_template.md:37]: Maybe add a little more detail, such as "(e.g. `main` or `release-name`)"

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+- [ ] Changes are limited to a single goal - *eschew scope creep!*
```
Comment (gdasher) [inline .github/pull_request_template.md:38]: Consider adding something about change being a manageable size.  Sometimes I've seen people commit really large changes that would likely be better broken into multiple smaller changes.  Even if there is a single goal, sometimes its better to do 2-3 small commits versus one big one, so long as the interim states are acceptable.

**Review by gdasher — APPROVED** [review]
LGTM

**Comment by gdasher** [comment]
Oops, sorry.  Accidentally hit the wrong button on a touchscreen and created a bad merge commit that you will need to revert.   Sorry!

**Review by nanda-katikaneni — APPROVED** [review]
Looks good to me - nothing else to add/change. Approved.

**Comment by tkol2022** [comment]
This is a nit pick. I suggest revising the line below to say "All unit tests and functional tests pass".

image
